pkg/controller/federatedcluster/controller.go

```diff
@@ -160,82 +150,127 @@ func (cc *ClusterController) Run(stopChan <-chan struct{}) {
-					if !(strings.EqualFold(string(clusterStatusNew.Conditions[i].Type), string(clusterStatusOld.Conditions[i].Type)) &&
-						strings.EqualFold(string(clusterStatusNew.Conditions[i].Status), string(clusterStatusOld.Conditions[i].Status))) {
-						hasTransition = true
-						break
-					}
-				}
-			}
+			store.resultRun = 1
 
-			if !hasTransition {
-				for j := 0; j < len(clusterStatusNew.Conditions); j++ {
-					clusterStatusNew.Conditions[j].LastTransitionTime = clusterStatusOld.Conditions[j].LastTransitionTime
+			probeTime := clusterStatusNew.Conditions[0].LastProbeTime
+			if isProbeSuccessful(clusterStatusNew) {
+				if store.resultRun < clusterHealthCheckConfig.SuccessThreshold {
+					// Success is below threshold - leave the probe state unchanged.
+					clusterStatusNew = store.clusterStatus
+					setLastProbeTime(clusterStatusNew, probeTime)
+				}
+			} else {
+				if store.resultRun < clusterHealthCheckConfig.FailureThreshold {
+					// failure is below threshold - leave the probe state unchanged.
+					clusterStatusNew = store.clusterStatus
+					setLastProbeTime(clusterStatusNew, probeTime)
 				}
 			}
 		}
+	}
+	store.clusterStatus = clusterStatusNew
 
-		if utilfeature.DefaultFeatureGate.Enabled(features.CrossClusterServiceDiscovery) {
-			zone, region, err := clusterClient.GetClusterZones()
-			if err != nil {
-				glog.Warningf("Failed to get zones and region for cluster %s: %v", cluster.Name, err)
-			} else {
+	if utilfeature.DefaultFeatureGate.Enabled(features.CrossClusterServiceDiscovery) {
+		zone, region, err := clusterClient.GetClusterZones()
+		if err != nil {
+			glog.Warningf("Failed to get zones and region for cluster %s: %v", cluster.Name, err)
+		} else {
+			// If new zone & region are empty, preserve the old ones so that user configured zone & region
+			// labels are effective
+			if store.clusterStatus != nil {
 				if len(zone) == 0 {
-					zone = cluster.Status.Zone
+					zone = store.clusterStatus.Zone
 				}
 				if len(region) == 0 {
-					region = cluster.Status.Region
+					region = store.clusterStatus.Region
 				}
-				clusterStatusNew.Zone = zone
-				clusterStatusNew.Region = region
 			}
+			clusterStatusNew.Zone = zone
+			clusterStatusNew.Region = region
 		}
+	}
 
-		cc.mu.Lock()
-		cc.clusterStatusMap[cluster.Name] = *clusterStatusNew
-		cc.mu.Unlock()
-		cluster.Status = *clusterStatusNew
-		err = cc.client.UpdateStatus(context.TODO(), &cluster)
-		if err != nil {
-			glog.Warningf("Failed to update the status of cluster: %v, error is : %v", cluster.Name, err)
-			// Don't return err here, as we want to continue processing remaining clusters.
-			continue
+	cluster.Status = *clusterStatusNew
+	if err := client.UpdateStatus(context.TODO(), cluster); err != nil {
+		glog.Warningf("Failed to update the status of cluster: %v, error is : %v", cluster.Name, err)
+	}
+	wg.Done()
+}
+
+func isProbeSuccessful(clusterStatus *fedv1a1.FederatedClusterStatus) bool {
```

Comment:
This should rather be named something like `isClusterOnline` or `isClusterAvailable`, isn't it.

Reply:
Have renamed to isClusterReady to reflect the cluster condition. PTAL.